### ISSUE TITLE
Support mounting floppy/CD images via drag and drop

### ISSUE THREAD

**Issue by dreamer** [dreamer/dosbox-staging#92]
Recently I installed Windows 3.11 inside dosbox-staging for testing, and nothing kills enthusiasm for installing older software then the necessity to juggle multiple floppy images. Especially if installer sometimes asks user to put in floppy out of order ("please insert this disk 5 out of 10").

It is totally possible to mount them alrady with `imgmount` and change with Ctrl+F4, but it is annoying due to lack of feedback.

With SDL2 it should be possible to implement mounting via [SDL_DropEvent](https://wiki.libsdl.org/SDL_DropEvent), which should make it much easier to mount floppies selectively.

Additionally, we could leave some visual indication about mounted floppy by placing e.g. `A:<label>` in the title bar (and remove "frameskip" information to make room).

**Comment by LowLevelMahn** [comment]
what about drag/drop of executables - maybe with auto-mount to exe-path?

**Comment by dreamer** [comment]
I see no reason why not :) It would be definitely very user-friendly feature.

**Comment by LowLevelMahn** [comment]
maybe too userfriendly:) 
console copy & paste is also very missing

**Comment by krcroft** [comment]
> Additionally, we could leave some visual indication about mounted floppy by placing e.g. `A:<label>` in the title bar (and remove "frameskip" information to make room).

This would be a big usability and feedback improvement! We could imgmount cue and  ISO files as well (or multiple.. looking at you Wing Command 4).

Indiication of the mount would be so nice to see; and if it replaces "frameskip" then all the better, which has outgrown its value with todays much faster computers.

**Comment by dreamer** [comment]
"Proof of concept" implementation can be found on the branch [`po/drag-drop-floppy-1`](https://github.com/dreamer/dosbox-staging/compare/po/drag-drop-floppy-1) (see snapshot builds in [attachments](https://github.com/dreamer/dosbox-staging/runs/368925456)).

Some cleanup, refactoring and proper implementation are needed, but this implementation is testable already.

- Works only for floppies (not for CD images nor executable files)
- Only 1 floppy image can be drag'n'dropped at a time
- Droping an image always targets drive A:
- Droping an image unmounts previous image mounted in drive A
- There is no visual feedback in the titlebar yet

Having briefly tested it… it's so nice to have DOSBox behaving like a user-friendly app for a while ;)

**Comment by krcroft** [comment]
Nice work @dreamer in laying the foundation for drag n' drop! 
(Looking forward to testing this when internet access to my home is restored)

**Comment by NicknineTheEagle** [comment]
Some kind of GUI for mounting is *badly* needed. I have all of my CD images stored in Redump naming convention which means that full paths are *long*. Having to type them out manually is insane, there needs to be a better way.

**Comment by NicknineTheEagle** [comment]
I would have *really* liked if this was addressed, this is the only issue that's stopping me from using dosbox-staging over upstream, it's otherwise superior in several ways.

**Comment by MasterO2** [comment]
@dreamer Any progress on this one?

**Comment by NicknineTheEagle** [comment]
Drag'n'drop alone is probably not the best solution, IMHO, DosBox-staging could use a proper GUI where you could mount images, change settings, etc. Dosbox-X actually has such a GUI although its configuration interface is pretty barebones.

**Comment by NicknineTheEagle** [comment]
Could we at least get auto-completion for imgmount command so that user doesn't have to type out the path manually?

**Comment by Grounded0** [comment]
Its probably up to your COMMAND.COM. I use FreeCOM from FreeDOS and it autocompletes nicely.

**Comment by NicknineTheEagle** [comment]
> Its probably up to your COMMAND.COM. I use FreeCOM from FreeDOS and it autocompletes nicely:
> 
> https://github.com/FDOS/freecom

No, you misunderstand, DosBox shell itself does support auto-completion but it only works on files withing DosBox's mounted filesystem, it does not work on host files.

**Comment by Burrito78** [comment]
@dreamer Do you still think this is doable under Linux/Win/macOS without hackery? Still in-scope? I would love to have it, btw!

**Comment by johnnovak** [comment]
Lots of weird comments here... Don't you use per-game configs with IMGMOUNT in your autoexec section, kids? 😄

**Comment by johnnovak** [comment]
Drag and drop is doable with floppy images. Multiple floppy images can be handled too.

With CD images, I think we need the OSD features otherwise selecting the drive letter would be too ambiguous. A requester dialog should come up, asking for the drive letter the CD image should be mounted to.

**Comment by Burrito78** [comment]
> Drag and drop is doable with floppy images. Multiple floppy images can be handled too.
> 
> With CD images, I think we need the OSD features otherwise selecting the drive letter would be too ambiguous. A requester dialog should come up, asking for the drive letter the CD image should be mounted to.

What about limiting mounting CD images via drag and drop to the next free drive letter after C? That is mostly correct on DOS machines and what most people will want. If you need something special, you have to mount via config/command line.

**Comment by johnnovak** [comment]
> > Drag and drop is doable with floppy images. Multiple floppy images can be handled too.
> > With CD images, I think we need the OSD features otherwise selecting the drive letter would be too ambiguous. A requester dialog should come up, asking for the drive letter the CD image should be mounted to.
> 
> What about limiting mounting CD images via drag and drop to the next free drive letter after C? That is mostly correct on DOS machines and what most people will want. If you need something special, you have to mount via config/command line.

I guess that could work too before the OSD is introduced. But the OSD requester approach is just superior IMO.

**Comment by johnnovak** [comment]
Good point, @Torinde, I'm convinced.

Probably this would be the best:

- Simple drag: mount file(s) as floppy image(s) in A:, or mount CD image(s) to the first available drive letter after C. If a CD-ROM mount already exists, swap the disc in the first CD-ROM drive instead.
- Shift + drag: Bring up requester with options.
